common/enclave/cc_data.cpp

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string>
+
+#include "base64.h"
+#include "cc_data.h"
+#include "error.h"
+#include "logging.h"
+
+#include <pb_decode.h>
+#include <pb_encode.h>
+#include "protos/fpc/attestation.pb.h"
+#include "protos/fpc/fpc.pb.h"
+
+#include "attestation-api/attestation/attestation.h"
+
+// ecc enclave global variable -- allocated dynamically
+cc_data* g_cc_data = NULL;
+
+bool cc_data::generate()
+{
+    return generate_keys();
+}
+
+bool cc_data::generate_keys()
+{
+    try
+    {
+        signature_key_.Generate();  // private key
+        verification_key_ = signature_key_.GetPublicKey();
+        decryption_key_.Generate();  // private key
+        encryption_key_ = decryption_key_.GetPublicKey();
+        cc_decryption_key_.Generate();  // private key
+        cc_encryption_key_ = cc_decryption_key_.GetPublicKey();
+
+        // debug
+        std::string s = verification_key_.Serialize();
+        LOG_DEBUG("enclave verification key: %s", s.c_str());
+    }
+    catch (...)
+    {
+        LOG_ERROR("error creating cryptographic keys");
+        return false;
+    }
+
+    return true;
+}
+
+bool cc_data::build_attested_data(ByteArray& attested_data)
+{
+    // estimate attested data size
+    // NOTE: the size is roughly estimated, it should be adapted
+    size_t attested_data_proto_size = attestation_parameters_.size() + cc_parameters_.size() +
+                                      host_parameters_.size() + (1 << 13);
+    uint8_t* attested_data_proto;
+    pb_ostream_t ostream;
+    bool b;
+
+    CATCH(b, attested_data.resize(attested_data_proto_size));
+    COND2ERR(!b);
+
+    attested_data_proto = attested_data.data();
+
+    ostream = pb_ostream_from_buffer(attested_data_proto, attested_data_proto_size);
+
+    {
+        // fpc_AttestedData_cc_params_tag
+        COND2ERR(!pb_encode_tag(&ostream, PB_WT_STRING, fpc_AttestedData_cc_params_tag));
+        COND2ERR(!pb_encode_string(
+            &ostream, (const unsigned char*)cc_parameters_.data(), cc_parameters_.size()));
+    }
+
+    {
+        // fpc_AttestedData_host_params_tag
+        COND2ERR(!pb_encode_tag(&ostream, PB_WT_STRING, fpc_AttestedData_host_params_tag));
+        COND2ERR(!pb_encode_string(
+            &ostream, (const unsigned char*)host_parameters_.data(), host_parameters_.size()));
+    }
+
+    {
+        // fpc_AttestedData_enclave_vk_tag
+        std::string s = verification_key_.Serialize();
```

Comment:
what is the format of this string? base64?

Reply:
In theory, it's opaque, meaning that the serialization should be consumed by the respective deserialize function.
Anyway, it's pem encoding.